<a href="https://colab.research.google.com/github/fabiohsst/LinkPredictionGNN/blob/main/lastfm_dataset_EDA.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
cd /content

/content


In [ ]:
!git clone https://github.com/fabiohsst/LinkPredictionGNN.git

Cloning into 'LinkPredictionGNN'...
remote: Enumerating objects: 4, done.
remote: Counting objects: 100% (4/4), done.
remote: Compressing objects: 100% (4/4), done.
remote: Total 4 (delta 0), reused 0 (delta 0), pack-reused 0
Receiving objects: 100% (4/4), done.


In [ ]:
# Data manipulation imports
import pandas as pd
import numpy as np

# plotly standard imports
import plotly.graph_objs as go
from plotly.offline import iplot
import matplotlib.pyplot as plt


In [ ]:
# Base settings for the plots

# 1. Colors
# Define a color palette
color_palette = {
    'primary': '#C8E4B2',
    'secondary': '#9ED2BE',
    'tertiary': '#7EAA92',
    'quaternary': '#FFD9B7',
    'background': '#fafafa',
    'text': '#333333'
}

# 2. Fonts
font_settings = {
    'family': 'Arial',
    'size': 14,
    'color': color_palette['text']
}

# 3. Layout and Design
layout_settings = {
    'plot_bgcolor': color_palette['background'],
    'paper_bgcolor': color_palette['background'],
    'font': font_settings,
    'margin': go.layout.Margin(l=50, r=50, b=50, t=50)
}

# Sample plot using the defined settings
sample_data = {
    'x': [1, 2, 3, 4],
    'y': [10, 11, 12, 13]
}

In [ ]:
# Defines columns name
column_names = ["userid", "timestamp", "artistid", "artist-name", "trackid", "track-name"]

# Load songs dataset
songs = pd.read_csv('/content/drive/MyDrive/Thesis/lastfm-dataset-1K/userid-timestamp-artid-artname-traid-traname.tsv',
                    sep='\t',names=column_names, error_bad_lines=False)

# Load users dataset
users = pd.read_csv('/content/drive/MyDrive/Thesis/lastfm-dataset-1K/userid-profile.tsv', sep='\t', error_bad_lines=False)

<ipython-input-7-be981359d56d>:5: FutureWarning:

The error_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.



Skipping line 2120260: expected 6 fields, saw 8

Skipping line 2446318: expected 6 fields, saw 8

Skipping line 11141081: expected 6 fields, saw 8

Skipping line 11152099: expected 6 fields, saw 12
Skipping line 11152402: expected 6 fields, saw 8

Skipping line 11882087: expected 6 fields, saw 8

Skipping line 12902539: expected 6 fields, saw 8
Skipping line 12935044: expected 6 fields, saw 8

Skipping line 17589539: expected 6 fields, saw 8

<ipython-input-7-be981359d56d>:9: FutureWarning:

The error_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.





In [ ]:
songs.head(10)

userid             timestamp                              artistid  \
0  user_000001  2009-05-04T23:08:57Z  f1b1cf71-bd35-4e99-8624-24a6e15f133a   
1  user_000001  2009-05-04T13:54:10Z  a7f7df4a-77d8-4f12-8acd-5c60c93f4de8   
2  user_000001  2009-05-04T13:52:04Z  a7f7df4a-77d8-4f12-8acd-5c60c93f4de8   
3  user_000001  2009-05-04T13:42:52Z  a7f7df4a-77d8-4f12-8acd-5c60c93f4de8   
4  user_000001  2009-05-04T13:42:11Z  a7f7df4a-77d8-4f12-8acd-5c60c93f4de8   
5  user_000001  2009-05-04T13:38:31Z  a7f7df4a-77d8-4f12-8acd-5c60c93f4de8   
6  user_000001  2009-05-04T13:33:28Z  a7f7df4a-77d8-4f12-8acd-5c60c93f4de8   
7  user_000001  2009-05-04T13:23:45Z  a7f7df4a-77d8-4f12-8acd-5c60c93f4de8   
8  user_000001  2009-05-04T13:19:22Z  a7f7df4a-77d8-4f12-8acd-5c60c93f4de8   
9  user_000001  2009-05-04T13:13:38Z  a7f7df4a-77d8-4f12-8acd-5c60c93f4de8   

  artist-name trackid                                  track-name  
0   Deep Dish     NaN  Fuck Me Im Famous (Pacha Ibiza)-09-28-2007  
1        坂本龍一     NaN           Composition 0919 (Live_2009_4_15)  
2        坂本龍一     NaN                        Mc2 (Live_2009_4_15)  
3        坂本龍一     NaN                     Hibari (Live_2009_4_15)  
4        坂本龍一     NaN                        Mc1 (Live_2009_4_15)  
5        坂本龍一     NaN                To Stanford (Live_2009_4_15)  
6        坂本龍一     NaN              Improvisation (Live_2009_4_15)  
7        坂本龍一     NaN                    Glacier (Live_2009_4_15)  
8        坂本龍一     NaN                  Parolibre (Live_2009_4_15)  
9        坂本龍一     NaN             Bibo No Aozora (Live_2009_4_15)

In [ ]:
users.head(10)

#id gender   age             country    registered
0  user_000001      m   NaN               Japan  Aug 13, 2006
1  user_000002      f   NaN                Peru  Feb 24, 2006
2  user_000003      m  22.0       United States  Oct 30, 2005
3  user_000004      f   NaN                 NaN  Apr 26, 2006
4  user_000005      m   NaN            Bulgaria  Jun 29, 2006
5  user_000006    NaN  24.0  Russian Federation  May 18, 2006
6  user_000007      f   NaN       United States  Jan 22, 2006
7  user_000008      m  23.0            Slovakia  Sep 28, 2006
8  user_000009      f  19.0       United States  Jan 13, 2007
9  user_000010      m  19.0              Poland   May 4, 2006

# Exploratory Data Analysis

## Dataset shape and missing values

### Songs

In [ ]:
# Print number of rows and columns of 'songs' dataset
rows, cols = songs.shape
print(f"The dataset 'songs' has {rows} rows and {cols} columns.")

The dataset 'songs' has 19098853 rows and 6 columns.


In [ ]:
# Missing values for each column
missing_values_songs = songs.isnull().sum()
missing_values_songs

userid               0
timestamp            0
artistid        600848
artist-name          0
trackid        2162719
track-name          12
dtype: int64

All the missing values in the 'songs' dataset will be deleted, mainly the id's values for artists and tracks, that will later be used as unique identifiers to the machine learning model.

In [ ]:
# Remove missing values on 'songs' dataset
songs.dropna(inplace=True)

In [ ]:
# Print number of rows and columns of 'songs' dataset after remove missing values
rows, cols = songs.shape
print(f"The dataset 'songs' has {rows} rows and {cols} columns.")

The dataset 'songs' has 16936134 rows and 6 columns.


In [ ]:
# Number of unique values for each column
unique_values_songs = songs.nunique()
unique_values_songs

userid              992
timestamp      15631466
artistid          83905
artist-name       81751
trackid          960402
track-name       693231
dtype: int64

### Users

In [ ]:
# Print number of rows and columns of 'users' dataset
rows, cols = users.shape
print(f"The dataset 'users' has {rows} rows and {cols} columns.")

The dataset 'users' has 992 rows and 5 columns.


In [ ]:
# Missing values for each column
missing_values_users = users.isnull().sum()
missing_values_users

#id             0
gender        108
age           706
country        85
registered      8
dtype: int64

Different of 'songs' dataset, there is no missing values on id's column. Furthermore, there are some on gender, age, country and registered columns. These missing values will not be deleted, although they will be valuable attributes later on the graph, it is possible to keep some of the nodes on the graph without them.

In [ ]:
# Number of unique values for each column
unique_values_users = users.nunique()
unique_values_users

#id           992
gender          2
age            34
country        66
registered    625
dtype: int64

## User's demopraphics

### Gender's distribution

In [ ]:
# Distribution of genders
gender_counts = users['gender'].value_counts()
total_users = len(users)

# Calculate percentages
male_percentage = (gender_counts.get('m', 0) / total_users) * 100
female_percentage = (gender_counts.get('f', 0) / total_users) * 100

# Create a Plotly bar chart
bar = go.Bar(x=gender_counts.index, y=gender_counts.values, name='Distribution of Genders',
             marker_color=[color_palette['primary'], color_palette['quaternary']],
             text=gender_counts.values, textposition='outside')

fig = go.Figure(data=[bar], layout=layout_settings)

# Update specific elements for this plot
fig.update_layout(
    title="<b>Gender distribution</b>",
    xaxis_title="<b>Gender</b>",
    yaxis_title="<b>Number of Users</b>"
)

# Show the figure
fig.show()

# Print the number and percentage of male and female users
print(f"Number of male users: {gender_counts.get('m', 0)} ({male_percentage:.2f}%)")
print(f"Number of female users: {gender_counts.get('f', 0)} ({female_percentage:.2f}%)")


Number of male users: 502 (50.60%)
Number of female users: 382 (38.51%)


 The gender of users is not equaly distributed, being the majority of Male (51.02%), while Female represents 38.82%. The remaining is from users that have no gender identified. It may create a uniqual accuracy on the model depending on the user gender.

### Age's distribution

In [ ]:
# Distribution of age
age_data = users['age'].dropna()

# Calculate histogram bins and frequencies
hist, bin_edges = np.histogram(age_data, bins=50)

# Create x-axis labels (midpoint of each bin)
bin_midpoints = (bin_edges[:-1] + bin_edges[1:]) / 2

# Create a bar chart that mimics a histogram
bar = go.Bar(x=bin_midpoints, y=hist, width=np.diff(bin_edges),
             marker_color=color_palette['primary'],
             text=hist,  # Adding text labels
             textposition='outside')  # Position of text labels

fig = go.Figure(data=[bar], layout=layout_settings)

# Update specific elements for this plot
fig.update_layout(
    title="<b>Age Distribution</b>",
    xaxis_title="<b>Age</b>",
    yaxis_title="<b>Number of Users</b>"
)

# Show the figure
fig.show()



In [ ]:
users['age'].describe()

count    286.000000
mean      25.367133
std        8.314952
min        3.000000
25%       21.000000
50%       23.000000
75%       28.000000
max      103.000000
Name: age, dtype: float64

The age of the users on the dataset is between 3 and 103 years, but the great majority are between 21 and 28 years old. The users on this range represents 75% of the users on the dataset. For this reason, the model might be more effective on this specific range. It is a point of attention for possible overfitting and small accuracy.

### Country's distribution

In [ ]:
# Countries distribution
top_countries = users['country'].value_counts().head(10).sort_values()
bar = go.Bar(
    y=top_countries.index,
    x=top_countries.values,
    orientation='h',
    marker_color=color_palette['primary'],
    text=top_countries.values,
    textposition='outside'
)
fig = go.Figure(data=[bar], layout=layout_settings)
fig.update_layout(
    title="<b>Top 10 Countries by User</b>",
    xaxis_title="<b>Number of Users</b>",
    yaxis_title="<b>Country</b>"
)
fig.show()


In [ ]:
# Total number of users
total_users = len(users)

# Number of users from United States and United Kingdom
us_users = len(users[users['country'] == 'United States'])
uk_users = len(users[users['country'] == 'United Kingdom'])

# Calculate percentages
us_percentage = (us_users / total_users) * 100
uk_percentage = (uk_users / total_users) * 100
rest_of_world_percentage = 100 - (us_percentage + uk_percentage)

n_countries = users['country'].nunique()

# Print the results
print(f"Number of countries: {n_countries}")
print(f"Percentage of users from United States: {us_percentage:.2f}%")
print(f"Percentage of users from United Kingdom: {uk_percentage:.2f}%")
print(f"Percentage of users from rest of the world: {rest_of_world_percentage:.2f}%")


Number of countries: 66
Percentage of users from United States: 22.98%
Percentage of users from United Kingdom: 12.70%
Percentage of users from rest of the world: 64.31%


 The users in the dataset are distributed on 66 countries. Even though, only two countries, United States and United Kingdom, represents 36% of the users. United States is the country with more users, with 23%, followed by United kingdom, with 13%. The dataset is evidently concentrated in two big markets.

### User's registration over time

In [ ]:
# Convert the 'registered' column to datetime format and coerce any errors
users['registered'] = pd.to_datetime(users['registered'], errors='coerce')

# Remove rows where 'registered' is NaT
users = users.dropna(subset=['registered'])

# Extracting month and year from the 'registered' column
users['month_year'] = users['registered'].dt.to_period('M')

# Grouping by month/year and counting registrations
registration_counts_monthly = users.groupby('month_year').size()

# Create a Plotly line chart
line = go.Scatter(x=registration_counts_monthly.index.astype(str),
                  y=registration_counts_monthly.values,
                  mode='lines+markers',
                  marker_color=color_palette['primary'])

fig = go.Figure(data=[line], layout=layout_settings)

# Update specific elements for this plot
fig.update_layout(
    title="<b>User Registrations Over Time (Monthly)</b>",
    xaxis_title="<b>Month/Year</b>",
    yaxis_title="<b>Number of Registrations</b>"
)

# Show the figure
fig.show()


<ipython-input-23-b1bcd1657c1c>:8: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [ ]:
# Extracting the year from the 'registered' column
users['year'] = users['registered'].dt.year

# Grouping by year and counting registrations
registration_counts_yearly = users.groupby('year').size()

# Convert the result to a DataFrame for better display
registration_yearly_df = registration_counts_yearly.reset_index()
registration_yearly_df.columns = ['Year', 'Number of Registrations']

# Display the table
display(registration_yearly_df)

<ipython-input-24-f706b503085a>:2: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



Year  Number of Registrations
0  2002                        1
1  2003                       13
2  2004                       77
3  2005                      302
4  2006                      438
5  2007                      147
6  2008                        2
7  2009                        4

The user registration column represents the date of sign up of the user to the platform. The most part of the registration was between 2004, with a shy increase in certain period, to a more agressive growth starting in 2005, with the pick in 2006 and a fast decrease in 2007. Beside this period, there a small number of users registred in 2002, 2003, 2008 and 2009.

### User listening habits

#### Distribution of number of songs per user

In [ ]:
# Distribution of number of songs per user
songs_per_user = songs['userid'].value_counts()

# Calculate histogram bins and frequencies
hist, bin_edges = np.histogram(songs_per_user, bins=30)

# Create x-axis labels (midpoint of each bin)
bin_midpoints = (bin_edges[:-1] + bin_edges[1:]) / 2

# Create a bar chart that mimics a histogram
bar = go.Bar(x=bin_midpoints, y=hist, width=np.diff(bin_edges),
             marker_color=color_palette['primary'],
             text=hist,
             textposition='outside')

fig = go.Figure(data=[bar], layout=layout_settings)

# Update specific elements for this plot
fig.update_layout(
    title="<b>Distribution of Number of Songs per User</b>",
    xaxis_title="<b>Number of Songs</b>",
    yaxis_title="<b>Number of Users</b>"
)

# Show the figure
fig.show()

In [ ]:
songs['userid'].value_counts().describe()

count       992.000000
mean      17072.715726
std       20800.361269
min           2.000000
25%        3351.500000
50%       10227.000000
75%       22294.000000
max      172042.000000
Name: userid, dtype: float64

The distribution in the number of songs listened for users is asymmetric with a positive skill. While the majority part of the users listen between 3,774 and 25,673 songs, there are user sample with 183,103 songs listened.

#### Distribution of Listening timestamp

In [ ]:
# Convert the 'timestamp' column to datetime format
songs['timestamp'] = pd.to_datetime(songs['timestamp'])

# Extract hour from timestamp
songs['hour'] = songs['timestamp'].dt.hour

In [ ]:
# Calculate histogram bins and frequencies
hist, bin_edges = np.histogram(songs['hour'], bins=24, range=(0, 24))

# Create x-axis labels (midpoint of each bin)
bin_midpoints = (bin_edges[:-1] + bin_edges[1:]) / 2

# Create a bar chart that mimics a histogram
bar = go.Bar(x=bin_midpoints, y=hist, width=np.diff(bin_edges),
             marker_color=color_palette['primary'],
             text=hist,
             textposition='outside')

fig = go.Figure(data=[bar], layout=layout_settings)

# Update specific elements for this plot
fig.update_layout(
    title="<b>Listening hours distribution</b>",
    xaxis_title="<b>Hour of Day</b>",
    yaxis_title="<b>Number of Songs Played</b>"
)

# Show the figure
fig.show()


In [ ]:
songs['hour'].describe().round()

count    16936134.0
mean           13.0
std             7.0
min             0.0
25%             8.0
50%            14.0
75%            19.0
max            23.0
Name: hour, dtype: float64

The time of each song listened is also recorded. This distribution is asymmetric with a negative skill. The number of songs listened starts increase significantly after 8:00 and still increasing until 18:00 where the number of listeners starts to decrease. This behaviour is observed generally saying, it is possible to investigate if deeper if this behaviour changes depending on the day of week and the country of the users.

## Songs and Artists Analysis

### Top artists by user interaction

In [ ]:
# Group by 'artistid', count the occurrences, and then sort the results
top_artist_ids = songs.groupby('artistid').size().sort_values(ascending=True).tail(10)

# Replace the 'artistid' index with 'artist-name' for the top artists
top_artist_names = top_artist_ids.index.map(lambda x: songs[songs['artistid'] == x]['artist-name'].iloc[0])
top_artist_ids.index = top_artist_names

# Create a Plotly horizontal bar chart
bar = go.Bar(y=top_artist_ids.index, x=top_artist_ids.values, orientation='h',
             marker_color=color_palette['primary'],
             text=top_artist_ids.values,
             textposition='outside')

fig = go.Figure(data=[bar], layout=layout_settings)

# Update specific elements for this plot
fig.update_layout(
    title="<b>Top 10 Artists by User Interactions</b>",
    xaxis_title="<b>Number of Interactions</b>",
    yaxis_title="<b>Artist Name</b>"
)

# Show the figure
fig.show()


In [ ]:
# Count the occurrences of each artist in the songs dataset
artist_counts = songs['artist-name'].value_counts()

# Convert the counts to a DataFrame for better display
artist_counts_df = artist_counts.reset_index()
artist_counts_df.columns = ['Artist', 'Number of Times Listened']

# Display the top artists from the table
display(artist_counts_df.head(10))


Artist  Number of Times Listened
0            Radiohead                    111488
1          The Beatles                     98525
2      Nine Inch Nails                     82022
3                 Muse                     60967
4             Coldplay                     60256
5         Depeche Mode                     57802
6  Death Cab For Cutie                     56451
7           Pink Floyd                     55739
8              Placebo                     50632
9        Elliott Smith                     49306

Analysing the artists, Radiohead and The Beatles are the most listed artists, followed by Nine Inch Nails, Muse and Coldplay. This dataset doesn't classify artists or songs by genre, so it is not possible to define the most popular genres. Even though, the five top artists are considered as rock bands.  

### Top songs by user interactions

In [ ]:
# Top songs by user interactions
top_songs = songs['track-name'].value_counts().head(10).sort_values(ascending=True)

# Create a Plotly horizontal bar chart
bar = go.Bar(y=top_songs.index, x=top_songs.values, orientation='h',
             marker_color=color_palette['primary'],
             text=top_songs.values,
             textposition='outside')

fig = go.Figure(data=[bar], layout=layout_settings)

# Update specific elements for this plot
fig.update_layout(
    title="<b>Top 10 Songs by User Interactions</b>",
    xaxis_title="<b>Number of Interactions</b>",
    yaxis_title="<b>Song Name</b>"
)

# Show the figure
fig.show()


In [ ]:
# Filtering the dataset for the song "Intro"
intro_song = songs[songs['track-name'] == 'Intro']

# 1. Number of unique users who listened to it
unique_users = intro_song['userid'].nunique()

# 2. Number of unique song id
unique_id = intro_song['trackid'].nunique()

# 3. Artists with the song "Intro"
artists_intro = intro_song['artist-name'].nunique()

print(f"Number of unique users who listened to 'Intro': {unique_users}")
print(f"Number of unique song id 'Intro'  : {unique_id}")
print(f"Number of unique artists who performed the song 'Intro': {artists_intro}")

Number of unique users who listened to 'Intro': 751
Number of unique song id 'Intro'  : 1400
Number of unique artists who performed the song 'Intro': 1400


In [ ]:
# Filtering the dataset for the song "[Untitled]"
untitled_1_song = songs[songs['track-name'] == '[Untitled]']

# 1. Number of unique users who listened to it
unique_users = untitled_1_song['userid'].nunique()

# 2. Number of unique song id
unique_id = untitled_1_song['trackid'].nunique()

# 3. Artists with the song "[Untitled]"
artists_intro = untitled_1_song['artist-name'].nunique()

print(f"Number of unique users who listened to '[Untitled]': {unique_users}")
print(f"Number of unique song id '[Untitled]'  : {unique_id}")
print(f"Number of unique artists who performed the song '[Untitled]': {artists_intro}")

Number of unique users who listened to '[Untitled]': 451
Number of unique song id '[Untitled]'  : 465
Number of unique artists who performed the song '[Untitled]': 465


First was printed the top 10 songs more listeds on the dataset and was verified that 'Intro' and '[Untitled]' were the most listed songs. Further investigation on these two songs were made to verify whether they represents unique songs from different artists or an error during record.
The song 'Intro' has 1400 unique ID's and was performed by 1400 different artists. While '[Untitled]' has 465 unique ID's and was performed by 465 artists. These are evidences that these two titles were correctly recorded and the fact of they be on the top 2 is because they are common names used in songs by many different artists.

In [ ]:
# Group by 'trackid', count the occurrences, and then sort the results
top_track_ids = songs.groupby('trackid').size().sort_values(ascending=True).tail(10)

# Replace the 'trackid' index with 'track-name' for the top songs
top_track_names = top_track_ids.index.map(lambda x: songs[songs['trackid'] == x]['track-name'].iloc[0])
top_track_ids.index = top_track_names

# Create a Plotly horizontal bar chart
bar = go.Bar(y=top_track_ids.index, x=top_track_ids.values, orientation='h',
             marker_color=color_palette['primary'],
             text=top_track_ids.values,
             textposition='outside')

fig = go.Figure(data=[bar], layout=layout_settings)

# Update specific elements for this plot
fig.update_layout(
    title="<b>Top Songs 10 by User Interactions (filtered)</b>",
    xaxis_title="<b>Number of Interactions</b>",
    yaxis_title="<b>Song Name</b>"
)

# Show the figure
fig.show()

When filtered the songs by their unique IDs - that is more accurate, as many songs can have the same name -  the top 5 songs changes.

### Popularity of songs over time

In [ ]:
# Assuming first timestamp as "release date"
song_release_date = songs.groupby('track-name')['timestamp'].min()
songs['song_release_date'] = songs['track-name'].map(song_release_date)
songs['days_since_release'] = (songs['timestamp'] - songs['song_release_date']).dt.days

In [ ]:
# Long longevity

# Calculate histogram bins and frequencies
hist, bin_edges = np.histogram(songs['days_since_release'].dropna(), bins=range(0, songs['days_since_release'].max() + 7, 7))

# Create x-axis labels (midpoint of each bin)
bin_midpoints = (bin_edges[:-1] + bin_edges[1:]) / 2

# Create a bar chart that mimics a histogram
bar = go.Bar(x=bin_midpoints, y=hist, width=np.diff(bin_edges),
             marker_color=color_palette['primary'])
fig = go.Figure(data=[bar], layout=layout_settings)

# Update specific elements for this plot
fig.update_layout(
    title="<b>Longevity of songs popularity</b>",
    xaxis_title="<b>Days Since Release (in bins of 7 days)</b>",
    yaxis_title="<b>Number of Records</b>",
    yaxis_type="log"  # Log scale
)

# Show the figure
fig.show()

In [ ]:
# Create bins for days_since_release in intervals of 7 days
bins = pd.cut(songs['days_since_release'], bins=range(0, songs['days_since_release'].max() + 7, 7))

# Count the number of records in each bin
days_since_release_counts = bins.value_counts().sort_index()

# Convert the result to a DataFrame for better display
days_since_release_df = days_since_release_counts.reset_index()
days_since_release_df.columns = ['Days Since Release (Interval)', 'Number of Records']

# Display the table
display(days_since_release_df)


Days Since Release (Interval)  Number of Records
0                          (0, 7]             299476
1                         (7, 14]             175270
2                        (14, 21]             139093
3                        (21, 28]             121020
4                        (28, 35]             112839
..                            ...                ...
391                  (2737, 2744]                  0
392                  (2744, 2751]                  0
393                  (2751, 2758]                  0
394                  (2758, 2765]                  0
395                  (2765, 2772]                  1

[396 rows x 2 columns]

In [ ]:
songs['days_since_release'].describe().round()

count    16936134.0
mean          635.0
std           443.0
min             0.0
25%           249.0
50%           603.0
75%           996.0
max          2771.0
Name: days_since_release, dtype: float64

The distribution of days since release aims to investigate if new songs are more popular among users than older and how long it last over time. There are no data from the release date of the songs. To execute this calcule was considered the first record of each song as the release data.
The result shows a tendency of the number of reproductions reduce significantly after the first week, and keep relatively stable after the first 200 days. After the day 1500 it seems to start decrease fast, although it might happen because of the data available. As the dataset goes from 2005 to 2009 (4 years) and the number of records available are not equal in every year. In addition, 1500 days is equivalent of 4.10 years. For this reasons is safier to assume that new songs are more popular in it first week and get less popular over the first trimester to later get relatively stable over the next years.

### Number of songs per artists

In [ ]:
# Artists with most unique songs
artists_song_counts = songs.groupby('artist-name')['trackid'].nunique().sort_values(ascending=True).tail(10)

# Create a Plotly horizontal bar chart
bar = go.Bar(y=artists_song_counts.index, x=artists_song_counts.values, orientation='h',
             marker_color=color_palette['primary'],
             text=artists_song_counts.values,
             textposition='outside')

fig = go.Figure(data=[bar], layout=layout_settings)

# Update specific elements for this plot
fig.update_layout(
    title="<b>Artists with Most Unique Songs</b>",
    xaxis_title="<b>Number of Unique Songs</b>",
    yaxis_title="<b>Artist Name</b>"
)

# Show the figure
fig.show()


Continuing the metadata analysis, the top 10 artists with more number of songs are relatively equal, with Johny Cash in first place with almost 1,000 unique songs on the dataset, followed by The Beatles, Depeche Mode, Johann Sebastian Bach and Frédéric Chopin with relatively the same number of songs.

## Temporal Analysis

### Listening trends over time

In [ ]:
songs.columns

Index(['userid', 'timestamp', 'artistid', 'artist-name', 'trackid',
       'track-name', 'hour', 'song_release_date', 'days_since_release'],
      dtype='object')

In [ ]:
# Group by date and count songs listened
songs_by_date = songs.groupby('timestamp').size()

# Create a Plotly line chart
line = go.Scatter(x=songs_by_date.index, y=songs_by_date.values, mode='lines',
                  line=dict(color=color_palette['primary']))

fig = go.Figure(data=[line], layout=layout_settings)

# Update specific elements for this plot
fig.update_layout(
    title="<b>Listening Trend Over Time</b>",
    xaxis_title="<b>Date</b>",
    yaxis_title="<b>Number of Songs Played</b>"
)

# Show the figure
fig.show()


In [ ]:
# Checking the distribution of samples per year
songs['year'] = songs['timestamp'].dt.year
songs_by_year = songs.groupby('year').size()
songs_by_year

The record of songs goes from 2005 to 2013. Even though there are only one record on the years 2010 and 2013. As it is not a significant representation it will me eliminated from the dataset.

In [ ]:
# Filtering out rows where the year in timestamp is 2010 or 2013
songs = songs[songs['timestamp'].dt.year.isin([2010, 2013]) == False]
songs_by_year = songs.groupby('year').size()
songs_by_year

### Artists trend over time

In [ ]:
# Group by date and artist, then count songs listened
songs_by_date_artist = songs.groupby(['timestamp', 'artist-name']).size().unstack().fillna(0)

# For demonstration, let's take the top 5 artists based on total listens and plot their trends
top_artists = songs['artist-name'].value_counts().head(5).index

# Initialize a list to store line traces
traces = []

# Create line traces for each artist using a rolling mean for smoother curves
for artist in top_artists:
    rolling_mean = songs_by_date_artist[artist].rolling(window=30).mean()
    trace = go.Scatter(x=rolling_mean.index, y=rolling_mean.values, mode='lines', name=artist)
    traces.append(trace)

# Create a Plotly figure
fig = go.Figure(data=traces, layout=layout_settings)

# Update specific elements for this plot
fig.update_layout(
    title="<b>Artists popularity over time (trend) Time</b>",
    xaxis_title="<b>Date</b>",
    yaxis_title="<b>Number of Songs Played</b>"
)

# Show the figure
fig.show()

The graph above represents the popularity (number of reproductions) of the top 5 artists over time. It is possible to see that Radiohead reached a pick of popularity in end 2007 that still decreasing until early 2008. The same happend to Nine Inch Nails in the middle of 2008. Beside these two picks, the popularity among these 5 artists flutuated over the years, with The beatles, Radiohead and Nine Inch Nails taking the first place in different moments.